### PR TITLE
Start of preValidateAdventure

### PR DESCRIPTION
**Pull request by Veracity0**
1) Since the portable spacegate (open) is its own zone, if we see the Spacegate in the mountains, we have permanent access.

2) Seeing the Barrel Shrine in the Dungeoneer's Association is permanent access

3) Seeing the Getaway Campsite in the woods is permanent access.

4) Adventure Validation part 0 - preValidateAdventure.
For things which have an expensive day pass (which we will not check/use in canAdventure/prepareForAdventure), if we don't have permanent access or know that we have daily access, take a peek at the place where the zone would appear, thereby setting daily access if appropriate.
- Spring Break Beach
- Conspiracy Island
- Dinsey Land Fill
- That 70's Volcano
- The Glaciest
- Gingerbread City
- Neverending Party
- Tunnel of L.O.V.E.
- The Spacegate

5) Unit tests for all of the above.

6) ASH function: boolean pre_validate_adventure(location)

7) My Sauceror got the Color Wheel of Yuck trophy and discovered that the vial of brown slime is multiusable.

### CODE REVIEW

File: src/net/sourceforge/kolmafia/KoLAdventure.java [hunk 1]
```diff
@@ -454,6 +481,143 @@ private void validate1() {
...
+      // Unlimited adventuring if available
+      if (Preferences.getBoolean("sleazeAirportAlways")) {
+        return true;
+      }
+      if (!Preferences.getBoolean("_sleazeAirportToday")) {
+        // Take a look at the airport.
+        var request = new PlaceRequest("airport");
+        RequestThread.postRequest(request);
+      }
+      return Preferences.getBoolean("_sleazeAirportToday");
```
Comment (midgleyc) [inline src/net/sourceforge/kolmafia/KoLAdventure.java:498]: All the airport versions are mostly the same, and could be refactored into `boolean checkAirportZone(String element)` to make the code read more easily.

You could go further and refactor everything but spacegate (and love tunnel, with its different comment) into `boolean checkZone(String alwaysPref, String todayPref, String place)`. I'm not sure whether this is better overall.

File: test/net/sourceforge/kolmafia/KoLAdventureValidationTest.java [hunk 2]
```diff
@@ -71,6 +73,213 @@ public static void afterAll() {
...
+      KoLAdventure adventure = AdventureDatabase.getAdventureByName(adventureName);
+      var html = html("request/test_visit_town_wrong.html");
+      // If we have always access, we don't have today access
+      checkDayPasses(adventure, "town_wrong", html, true, false, always, today);
+      // If we don't have always access, we might today access
+      checkDayPasses(adventure, "town_wrong", html, false, true, always, today);
+      // If we don't have always or today access, we might still have today access
+      checkDayPasses(adventure, "town_wrong", html, false, false, always, today);
+      // If we don't have always or today access, we might really not have today access
+      checkDayPasses(adventure, "town_wrong", null, false, false, always, today);
```
Comment (midgleyc) [inline test/net/sourceforge/kolmafia/KoLAdventureValidationTest.java:138]: This block of code is the same bar strings and could be refactored into `void checkPassCombinations(String place)`.

**Review by midgleyc — CHANGES_REQUESTED** [review]
Two small function extractions I think are beneficial.

File: src/net/sourceforge/kolmafia/KoLAdventure.java (hunk 1 above)
Comment (Veracity0) [reply src/net/sourceforge/kolmafia/KoLAdventure.java:498]: checkZone works well and is clean.

Comment (Veracity0) [reply src/net/sourceforge/kolmafia/KoLAdventure.java]: Done.

File: test/net/sourceforge/kolmafia/KoLAdventureValidationTest.java (hunk 2 above)
Comment (Veracity0) [reply test/net/sourceforge/kolmafia/KoLAdventureValidationTest.java]: Done.

**Review by midgleyc — APPROVED** [review]

**Pull request merged**